onnx/onnx.proto

```diff
@@ -53,27 +53,49 @@ enum Version {
   // control. We should use version as
   //     xx(major) - xx(minor) - xxxx(bugfix)
   // and we are starting with 00000001.
-  IR_VERSION = 00000001;
+  IR_VERSION = 00000101;
+  IR_VERSION_2017_10_10 = 00000001;
 }
 
+
+
 // A named attribute containing either singular float, integer, string
 // and tensor values, or repeated float, integer, string and tensor values.
 // An AttributeProto MUST contain the name field, and *only one* of the
 // following content fields, effectively enforcing a C/C++ union equivalent.
+
+enum AttributeType {
+  STRING = 0;
+  FLOAT = 1;
+  INT = 2;
+  TENSOR = 3;
+  GRAPH = 4;
+
+  STRINGS = 5;
+  FLOATS = 6;
+  INTS = 7;
+  TENSORS = 8;
+  GRAPHS = 9;
+}
+
 message AttributeProto {
   // The name field MUST be present for this version of the IR.
   optional string name = 1;           // namespace Attribute
-  optional float f = 2;               // float
-  optional int64 i = 3;               // int
-  optional bytes s = 4;               // UTF-8 string
-  optional TensorProto t = 5;         // tensor value
-  optional GraphProto g = 6;          // graph
-
-  repeated float floats = 7;          // list of floats
-  repeated int64 ints = 8;            // list of ints
-  repeated bytes strings = 9;         // list of UTF-8 strings
-  repeated TensorProto tensors = 10;  // list of tensors
-  repeated GraphProto graphs = 11;    // list of graph
+  // The type field MUST be present for this version of the IR.
```

Comment:
even with BC we shouldn't break proto compatibility (i.e. be able to deserialize). Let's keep the ids for existing fields.

Reply:
Good point. 

I think what you are advocating is this:
    1. Allow pre-1.1 implementations to simply infer the discriminator value for 1.0 files.
    2. Require 1.1 or later files to have the field written.  Absent (which defaults to 0 which is FLOAT) would always be treated as a FLOAT and any has_field heuristics would be illegal.

If we do this, then technically this isn't a breaking change (which I like!).

Given that, I'd suggest we simply bump the bugfix number in the version.